amcl:
  ros__parameters:
    use_sim_time: true
    alpha1: 0.2
    alpha2: 0.2
    alpha3: 0.2
    alpha4: 0.2
    alpha5: 0.2
    base_frame_id: base_link
    odom_frame_id: odom
    global_frame_id: map
    scan_topic: /lidar_sensor_plugin/out
    min_particles: 500
    max_particles: 2000
    set_initial_pose: true
    save_pose_rate: 0.5
    update_min_d: 0.25
    update_min_a: 0.2
    resample_interval: 1
    transform_tolerance: 1.0
    initial_pose: {x: 0.0, y: 0.0, yaw: 0.0}

map_server:
  ros__parameters:
    use_sim_time: true
    yaml_filename: maps/Hallway_map.yaml

planner_server:
  ros__parameters:
    expected_planner_frequency: 20.0
    planner_plugins: ["GridBased"]
    GridBased:
      plugin: "nav2_navfn_planner/NavfnPlanner"
      tolerance: 0.5
      use_astar: false
      allow_unknown: true

controller_server:
  ros__parameters:
    use_sim_time: true
    controller_frequency: 20.0
    min_x_velocity: -0.3
    max_x_velocity: 0.5
    min_y_velocity: 0.0
    max_y_velocity: 0.0
    min_theta_velocity: -1.0
    max_theta_velocity: 1.0
    acceleration_limits: [1.5, 0.0, 3.2]  # [x, y, theta]
    deceleration_limits: [1.5, 0.0, 3.2]
    controller_plugins: ["FollowPath"]

    FollowPath:
      plugin: "dwb_core::DWBLocalPlanner"
      debug_trajectory_details: true
      min_vel_x: 0.0
      max_vel_x: 0.5
      min_vel_y: 0.0
      max_vel_y: 0.0
      max_vel_theta: 1.0
      min_speed_xy: 0.0
      max_speed_xy: 0.5
      min_speed_theta: 0.0
      acc_lim_x: 1.5
      acc_lim_y: 0.0
      acc_lim_theta: 3.2
      decel_lim_x: -1.5
      decel_lim_y: 0.0
      decel_lim_theta: -3.2
      vx_samples: 20
      vy_samples: 0
      vtheta_samples: 20
      sim_time: 2.0
      linear_granularity: 0.05
      angular_granularity: 0.025
      transform_tolerance: 0.2
      xy_goal_tolerance: 0.25
      yaw_goal_tolerance: 0.25
      critics: ["RotateToGoal", "Oscillation", "ObstacleFootprint", "PathAlign", "PathDist", "GoalDist"]
      ObstacleFootprint.scale: 0.02
      PathAlign.scale: 32.0
      PathDist.scale: 32.0
      GoalDist.scale: 24.0

bt_navigator:
  ros__parameters:
    use_sim_time: true
    default_bt_xml_filename: "navigate_w_replanning_and_recovery.xml"
    global_frame: map
    robot_base_frame: base_link
    odom_topic: odom
    transform_tolerance: 0.2

waypoint_follower:
  ros__parameters:
    use_sim_time: true
    stop_on_failure: true
    loop_rate: 20

recoveries_server:
  ros__parameters:
    costmap_topic: global_costmap/costmap_raw
    footprint_topic: global_costmap/published_footprint
    cycle_frequency: 10.0
    recovery_plugins: ["Spin", "BackUp", "Wait"]
    Spin:
      plugin: "nav2_recoveries/Spin"
    BackUp:
      plugin: "nav2_recoveries/BackUp"
    Wait:
      plugin: "nav2_recoveries/Wait"

global_costmap:
  global_costmap:
    ros__parameters:
      use_sim_time: true
      robot_base_frame: base_link
      global_frame: map
      update_frequency: 5.0
      publish_frequency: 2.0
      resolution: 0.05
      width: 50
      height: 50
      footprint: "[[0.25, 0.2], [0.25, -0.2], [-0.25, -0.2], [-0.25, 0.2]]"
      plugins: ["static_layer", "obstacle_layer", "inflation_layer"]

      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: true

      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        observation_sources: scan
        scan:
          topic: /lidar_sensor_plugin/out
          max_obstacle_height: 2.0
          clearing: true
          marking: true

      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        inflation_radius: 0.55
        cost_scaling_factor: 10.0

local_costmap:
  local_costmap:
    ros__parameters:
      use_sim_time: true
      robot_base_frame: base_link
      global_frame: odom
      update_frequency: 10.0
      publish_frequency: 2.0
      width: 10
      height: 10
      resolution: 0.05
      rolling_window: true
      footprint: "[[0.25, 0.2], [0.25, -0.2], [-0.25, -0.2], [-0.25, 0.2]]"
      plugins: ["obstacle_layer", "inflation_layer"]

      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        observation_sources: scan
        scan:
          topic: /lidar_sensor_plugin/out
          max_obstacle_height: 2.0
          clearing: true
          marking: true

      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        inflation_radius: 0.55
        cost_scaling_factor: 10.0

lifecycle_manager:
  ros__parameters:
    use_sim_time: true
    autostart: true
    node_names: [
      'map_server',
      'amcl',
      'planner_server',
      'controller_server',
      'recoveries_server',
      'bt_navigator',
      'waypoint_follower'
    ]